### PR TITLE
Initial commit of zapi cli with info command

### PR DESCRIPTION
**Pull request by mattnibs**
Adds a barebones version of zapi: A simple cli for interacting with
zqd servers.

To reduce burden on reviewers the scope has been limited to and single
command: info (additionally the stock charm help command also works).
Future functionality will eventually cover the entirety of the zqd
api including the ability to create, update, remove, post data to
spaces.

The info commands allows user to view information on space(s). The
space(s) can be queried by using either the global space selection
flags (i.e. -n or -id) or by use one or more globbed enhanced arguments.

part of #526

### CODE REVIEW

**Comment by mattnibs** [comment]
Going to look how difficult it is to add a few rudimentary tests with ztest.

File: cmd/zapi/cmd/reflect.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+		if JsonName(t.Field(i)) == name {
+			return v.Field(i), true
+		}
+	}
+	return reflect.Value{}, false
+}
+
+// JsonName returns the JSON name of the field.  It returns the empty string if
+// the field is always omitted (i.e., it has a json:"-" tag).
+func JsonName(s reflect.StructField) string {
```
Comment (nwt) [inline cmd/zapi/cmd/reflect.go:22]:
```suggestion
func JSONName(s reflect.StructField) string {
```

File: cmd/zapi/cmd/cli.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+var (
+	// version numbers set by main
+	Version   string
+	ZqVersion string
+)
+
+var ErrSpaceNotSpecified = errors.New("either space name (-s) or id (-id) must be specified")
+
+var Cli = &charm.Spec{
```
Comment (nwt) [inline cmd/zapi/cmd/cli.go:23]:
```suggestion
var CLI = &charm.Spec{
```

File: cmd/zapi/cmd/cli.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+	Cli.Add(charm.Help)
+}
+
+func New(f *flag.FlagSet) (charm.Command, error) {
+	c := &Command{
+		Version:   Version,
+		ZqVersion: ZqVersion,
+	}
+
+	defaultHost := "localhost:9867" //XXX
```
[the remark this answers is not in the document]
Comment (mattnibs) [reply cmd/zapi/cmd/cli.go:44]: This is copy pasta- I will remove.

File: zqd/api/api.go [hunk 4]
```diff
@@ -72,8 +74,26 @@ type ScannerStats struct {
...
 
+var spaceIDRegexp = regexp.MustCompile("^[a-zA-Z0-9_]+$")
+
 type SpaceID string
 
+func (s SpaceID) String() string {
+	return string(s)
+}
+
+func (s SpaceID) Get() interface{} {
```
[the remark this answers is not in the document]
Comment (mattnibs) [reply zqd/api/api.go:85]: Yep, good idea.

File: cmd/zapi/cmd/api.go [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+		url.Host += fmt.Sprintf(":%d", api.DefaultPort)
+	}
+	c := api.NewConnectionTo(url.String())
+	c.SetUserAgent("ZQD-CLI")
+	return &API{c}, nil
+}
+
+func (a API) Native() *api.Connection {
+	return a.Connection
+}
```
[the remark this answers is not in the document]
Comment (mattnibs) [reply cmd/zapi/cmd/api.go:35]: I was thinking of getting rid of this. api had one useful override method, `SpaceExists` but I was thinking of moving this to connection or to the root command.

File: cmd/zapi/cmd/cli.go [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+	Version   string
+	ZqVersion string
+	Host      string
+	Spacename string
+	spaceID   api.SpaceID
+}
+
+// API returns the api object.  If it doesn't exist, it is allocated and
+// the server is contacted and authenticated.  If the user types in a new
+// password to auhenticate, then the password is saved in the credentials file.
```
[the remark this answers is not in the document]
Comment (mattnibs) [reply cmd/zapi/cmd/cli.go:63]: Yep, will remove.

File: cmd/zapi/cmd/cli.go [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+	ZqVersion string
+)
+
+var ErrSpaceNotSpecified = errors.New("either space name (-s) or id (-id) must be specified")
+
+var Cli = &charm.Spec{
+	Name:          "zapi",
+	Usage:         "zapi [global options] command [options] [arguments...]",
+	Short:         "use zapi to talk to a zqd server",
+	RedactedFlags: "p",
```
[the remark this answers is not in the document]
Comment (mattnibs) [reply cmd/zapi/cmd/cli.go:27]: I'm not rightly sure. I've removed this (sorry I had thought I removed all boom cli artifacts).

**Pull request merged**